core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -331,17 +418,22 @@ class LogManager(val logDirs: Array[File],
           log.truncateTo(truncateOffset)
           if (needToStopCleaner)
             cleaner.maybeTruncateCheckpoint(log.dir.getParentFile, topicPartition, log.activeSegment.baseOffset)
+        } catch {
+          case e: IOException =>
+            error(s"Failed to truncate log for $topicPartition due to IOException", e)
+            handleLogDirFailure(log.dir.getParent)
```

Comment:
Yeah I would like to invoke `ReplicaManager.handleLogDirFailure()` whenever there is IOException. But this is not done yet because `LogManager` is not able to reference `ReplicaManager` since it is instantiated before the `ReplicaManager`. We can address this problem by doing `LogManager.setOnLogDirFailureCallback()` after the ReplicaManage is instantiated. I find this to be a big ugly though. What do you think?